<launch>

	<arg name="bagfile" default="<!-- Your bagfile here -->"/>

	<!-- Setup the camera -->
	<arg name="camera" default="/camera"/>
	<param name="/use_sim_time" value="true"/>

	<!-- Play the bagfile -->
	<node pkg="rosbag" type="play" name="rosbag" args="--clock $(arg bagfile)"/>

	<!-- Run the stereo image proc -->
	<group ns="$(arg camera)">
		<node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc" />
	</group>

	<!-- Run the visual odometry: You can choose between viso2 and fovis -->
	<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer">
		<remap from="stereo" to="$(arg camera)"/>
		<remap from="image" to="image_rect_color"/>
		<param name="base_link_frame_id" value="$(arg camera)"/>
	</node>
	<!--node pkg="fovis_ros" type="stereo_odometer" name="fovis">
		<remap from="stereo" to="$(arg camera)"/>
		<remap from="image" to="image_rect_color"/>
		<param name="base_link_frame_id" value="$(arg camera)"/>
	</node-->

	<!-- Stereo SLAM -->
	<node pkg="stereo_slam" type="slam" name="slam" output="screen">

		<!-- Working directory -->
		<param name="work_dir" value="<!-- Path to your data directory (e.g. /home/user/stereo_slam/data) -->" />

		<!-- Topic parameters -->
		<param name="odom_topic" value="stereo_odometer/odometry"/>
		<param name="left_topic" value="$(arg camera)/left/image_rect_color"/>
		<param name="right_topic" value="$(arg camera)/right/image_rect_color"/>
		<param name="left_info_topic" value="$(arg camera)/left/camera_info"/>
		<param name="right_info_topic" value="$(arg camera)/right/camera_info"/>

		<!-- Min displacement between graph vertices (in meters) -->
		<param name="min_displacement" value="0.25"/>

		<!-- Descriptor type: SIFT or SURF -->
		<param name="desc_type" value="SIFT"/>

		<!-- Descriptor threshold (for SIFT typically between 0.8-0.9)  -->
		<param name="desc_thresh" value="0.88" />

		<!-- Minimum number of neighbours that will be skiped for the loop closure (tipically between 5-20, but depends on the frame rate) -->
		<param name="min_neighbour" value="10" />

		<!-- Get the n first candidates of the hash matching (tipically between 1-5) -->
		<param name="n_candidates" value="5" />

		<!-- Minimun number of descriptor matches to consider a matching as possible loop closure (>8) -->
		<param name="min_matches" value="30" />

		<!-- Minimum number of inliers to consider a matching as possible loop closure (>8) -->
		<param name="min_inliers" value="15" />

		<!-- Set to 0 for LinearSlam Solver with gauss-newton. Set to 1 for LinearSlam Solver with Levenberg -->
		<param name="g2o_algorithm" value="0" />

		<!-- Maximum number of iteration for the g2o algorithm -->
		<param name="g2o_opt_max_iter" value="10" />
	</node>

	<!-- Pointcloud collector -->
	<node pkg="stereo_slam" type="collector" name="collector" output="screen">

		<param name="work_dir" value="<!-- Path to your data directory (e.g. /home/user/stereo_slam/data) -->" />

		<param name="graph_topic" value="slam/graph"/>
		<param name="cloud_topic" value="$(arg camera)/scaled/points2"/>

		<rosparam file="$(find stereo_slam)/etc/reconstruction.yaml" />

	</node>

</launch>
